File: speech_to_text/csharp/transcribe_file/Remote/Remote.cs
using System;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Threading.Tasks;

string apiBase = "https://api.soniox.com";
string audioUrl = "https://soniox.com/media/examples/coffee_shop.mp3";

// Retrieve the API key from environment variable (ensure SONIOX_API_KEY is set)
string? apiKey = Environment.GetEnvironmentVariable("SONIOX_API_KEY");
if (string.IsNullOrEmpty(apiKey))
{
    throw new InvalidOperationException("SONIOX_API_KEY is not set.");
}

// Create a requests session and set the Authorization header
var client = new HttpClient();
client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", apiKey);

// 1. Start a new transcription session by sending the audio URL to the API
Console.WriteLine("Starting transcription...");

var request = new StringContent(
    JsonSerializer.Serialize(new { audio_url = audioUrl, model = "stt-async-preview" }),
    System.Text.Encoding.UTF8,
    "application/json"
);
var response = await client.PostAsync($"{apiBase}/v1/transcriptions", request);
response.EnsureSuccessStatusCode();
var responseJson = await response.Content.ReadAsStringAsync();

var transcription = JsonNode.Parse(responseJson);
var transcriptionId = transcription?["id"]?.ToString();
if (string.IsNullOrEmpty(transcriptionId))
{
    throw new InvalidOperationException("Transcription ID is missing.");
}

Console.WriteLine($"Transcription started with ID: {transcriptionId}");

// 2. Poll the transcription endpoint until the status is 'completed'
while (true)
{
    response = await client.GetAsync($"{apiBase}/v1/transcriptions/{transcriptionId}");
    response.EnsureSuccessStatusCode();
    responseJson = await response.Content.ReadAsStringAsync();
    transcription = JsonNode.Parse(responseJson);

    var status = transcription?["status"]?.ToString();
    if (status == "error")
    {
        var errorMessage = transcription?["error_message"]?.ToString() ?? "Unknown error";
        throw new Exception($"Transcription error: {errorMessage}");
    }
    else if (status == "completed")
    {
        break;
    }

    await Task.Delay(1000);
}

// 3. Retrieve the final transcript once transcription is completed
response = await client.GetAsync($"{apiBase}/v1/transcriptions/{transcriptionId}/transcript");
response.EnsureSuccessStatusCode();
responseJson = await response.Content.ReadAsStringAsync();

var transcript = JsonNode.Parse(responseJson);
var transcriptText = transcript?["text"]?.ToString();
if (string.IsNullOrEmpty(transcriptText))
{
    throw new InvalidOperationException("Transcript text is missing.");
}

Console.WriteLine("Transcript:");
Console.WriteLine(transcriptText);

File: speech_to_text/java/transcribe_file/remote/src/main/java/com/soniox/example/remote/Remote.java
package com.soniox.example.remote;

import java.io.File;
import java.io.IOException;

import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.ContentType;
import org.apache.http.entity.mime.MultipartEntityBuilder;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.util.EntityUtils;
import org.json.JSONObject;

public class Remote {
    public static void main(String[] args) throws Exception {
        String apiBase = "https://api.soniox.com";
        String apiKey = System.getenv("SONIOX_API_KEY");

        try (CloseableHttpClient client = HttpClients.createDefault()) {

            // 1. Create a transcription
            System.out.println("Starting transcription...");

            JSONObject createTranscriptionBody = new JSONObject();
            createTranscriptionBody.put("audio_url", "https://soniox.com/media/examples/coffee_shop.mp3");
            createTranscriptionBody.put("model", "stt-async-preview");

            HttpPost createTranscriptionReq = new HttpPost(apiBase + "/v1/transcriptions");
            createTranscriptionReq.setHeader("Authorization", "Bearer " + apiKey);
            createTranscriptionReq.setHeader("Content-Type", "application/json");
            createTranscriptionReq.setEntity(new StringEntity(createTranscriptionBody.toString(), ContentType.APPLICATION_JSON));


            HttpResponse res = client.execute(createTranscriptionReq);
            if (res.getStatusLine().getStatusCode() != 201) {
                throw new RuntimeException("Failed to create transcription: " + EntityUtils.toString(res.getEntity()));
            }

            JSONObject transcription = new JSONObject(EntityUtils.toString(res.getEntity()));

            String transcriptionId = transcription.getString("id");

            System.out.println("Transcription started with ID: " + transcriptionId);

            // Wait for the transcription to complete

            while (true) {
                HttpGet transcriptionReq = new HttpGet(apiBase + "/v1/transcriptions/" + transcriptionId);
                transcriptionReq.setHeader("Authorization", "Bearer " + apiKey);

                res = client.execute(transcriptionReq);
                if (res.getStatusLine().getStatusCode() != 200) {
                    throw new RuntimeException("Failed to get transcription: " + EntityUtils.toString(res.getEntity()));
                }

                transcription = new JSONObject(EntityUtils.toString(res.getEntity()));

                if ("error".equals(transcription.getString("status"))) {
                    throw new RuntimeException("Transcription error: " + transcription.getString("error_message"));
                } else if ("completed".equals(transcription.getString("status"))) {
                    break;
                }

                // Wait for 1 second before polling again
                Thread.sleep(1000);
            }

            // Get the transcript

            HttpGet transcriptReq = new HttpGet(apiBase + "/v1/transcriptions/" + transcription.getString("id") + "/transcript");
            transcriptReq.setHeader("Authorization", "Bearer " + apiKey);

            res = client.execute(transcriptReq);
            if (res.getStatusLine().getStatusCode() != 200) {
                throw new RuntimeException("Failed to get transcript: " +
                        EntityUtils.toString(res.getEntity()));
            }

            JSONObject transcript = new JSONObject(EntityUtils.toString(res.getEntity()));

            System.out.println("Transcript:");
            System.out.println(transcript.getString("text"));
        }
    }
}
